src/socket/communicate.rs

```diff
@@ -0,0 +1,239 @@
+//! Communication with the lorri daemon.
+//!
+//! We provide a fixed number of `CommunicationType`s that are sent
+//! between the lorri daemon and lorri clients wishing to connect to it.
+//!
+//! `listener` implements the daemon side, which provides an `accept`
+//! method which is similar to how `accept()` works on a plain Unix socket.
+//!
+//! `client` implements a set of clients specialized to the communications
+//! we support.
+
+use std::os::unix::net::UnixStream;
+
+use crate::socket::path::{BindError, BindLock, SocketPath};
+use crate::socket::{ReadWriteError, ReadWriter, Timeout};
+use crate::NixFile;
+
+/// We declare 1s as the time readers should wait
+/// for the other side to send something.
+pub const DEFAULT_READ_TIMEOUT: Timeout = Timeout::from_millis(1000);
+
+/// Enum of all communication modes the lorri daemon supports.
+#[derive(Serialize, Deserialize)]
+pub enum CommunicationType {
+    /// Ping the daemon from a project to tell it to watch & evaluate
+    // TODO: rename to BuildProject or something?
+    Ping,
+    // /// Listen for events on the daemon
+    // DaemonListener,
+    // /// Get the daemon status
+    // DaemonStatus,
+}
+
+/// Message sent by the client to ask the server to start
+/// watching `nix_file`. See `CommunicationType::Ping`.
+#[derive(Serialize, Deserialize)]
+pub struct Ping {
+    /// The nix file to watch and build on changes.
+    pub nix_file: NixFile,
+}
+
+/// No message can be sent through this socket end (empty type).
+pub enum NoMessage {}
+// pub struct DaemonEvent {}
+// pub struct DaemonStatus {}
+
+/// `Listener` and possible errors.
+pub mod listener {
+    use super::*;
+    use std::os::unix::net::{UnixListener, UnixStream};
+
+    /// If a connection on the socket is attempted and the first
+    /// message is of a `ConnectionType`, the `Listener` returns
+    /// this message as an ack.
+    /// In all other cases the `Listener` returns no answer (the
+    /// bad client should time out after some time).
+    #[derive(Debug, Serialize, Deserialize)]
+    pub struct ConnectionAccepted();
+
+    /// Server-side part of a socket transmission,
+    /// listening for incoming messages.
+    pub struct Listener {
+        /// Bound Unix socket.
+        listener: UnixListener,
+        /// Lock that keeps our socket exclusive.
+        #[allow(dead_code)]
```

Comment:
What is this `allow` about?

Reply:
`bind_lock` is never explicitely used anywhere, the lock is dropped when `Listener` goes out of scope, so we can hide the warning.